File: web/src/patch.ts
/**
 * JSON Patch (RFC 6902) — 서버가 보낸 연산을 문서에 적용한다.
 *
 * 서버 쪽 `core/patch.py`와 **같은 연산 넷**을 구현한다. 둘이 같은 규격을
 * 보고 있어서 양쪽이 서로의 구현을 몰라도 된다. 그것이 규격을 쓰는 이유다.
 *
 * 불변으로 만든다(구조 공유 복사). React가 바뀐 가지만 다시 그리게 하려면
 * 제자리에서 고치면 안 된다.
 */
export type Op =
  | { op: 'add'; path: string; value: unknown }
  | { op: 'replace'; path: string; value: unknown }
  | { op: 'remove'; path: string }
  | { op: 'move'; from: string; path: string }

const segments = (path: string) =>
  path.split('/').slice(1).map(s => s.replace(/~1/g, '/').replace(/~0/g, '~'))

function clone<T>(value: T): T {
  if (Array.isArray(value)) return [...value] as unknown as T
  if (value && typeof value === 'object') return { ...value } as T
  return value
}

/** 경로를 따라 내려가며 지나는 노드만 복사한다. 나머지 가지는 그대로 공유한다. */
function copyPath(root: any, parts: string[]): [any, any, string] {
  const next = clone(root)
  let cursor = next
  for (let i = 0; i < parts.length - 1; i += 1) {
    const key = Array.isArray(cursor) ? Number(parts[i]) : parts[i]
    cursor[key] = clone(cursor[key])
    cursor = cursor[key]
  }
  return [next, cursor, parts[parts.length - 1]]
}

function getAt(root: any, path: string): unknown {
  return segments(path).reduce((cur, key) => (cur == null ? cur : cur[key as any]), root)
}

export function applyOne<T>(doc: T, op: Op): T {
  if (op.op === 'move') {
    const value = getAt(doc, op.from)
    const removed = applyOne(doc, { op: 'remove', path: op.from })
    return applyOne(removed, { op: 'add', path: op.path, value })
  }

  const [next, parent, key] = copyPath(doc, segments(op.path))

  if (op.op === 'add') {
    if (Array.isArray(parent)) {
      key === '-' ? parent.push(op.value) : parent.splice(Number(key), 0, op.value)
    } else {
      parent[key] = op.value
    }
  } else if (op.op === 'replace') {
    parent[Array.isArray(parent) ? Number(key) : (key as any)] = op.value
  } else if (op.op === 'remove') {
    Array.isArray(parent) ? parent.splice(Number(key), 1) : delete parent[key]
  }
  return next
}

export function applyPatch<T>(doc: T, ops: Op[]): T {
  return ops.reduce<T>((current, op) => applyOne(current, op), doc)
}

/** 바뀐 자리에 잠깐 불을 켜려고 쓴다.
 *
 *  `/sections/2/rows` → 섹션 2, `/period/from` → 머리말.
 *  **패치가 어디를 고쳤는지 아는 것이 화면을 덜 깜빡이게 하는 값이다.**
 *  스냅샷을 받으면 이것을 알 수 없어 전부 다시 그리게 된다. */
export type Touched = { sections: Set<number>; head: boolean; conclusion: boolean }

export function touched(ops: Op[]): Touched {
  const out: Touched = { sections: new Set(), head: false, conclusion: false }
  for (const op of ops) {
    if (op.path === '/sections') out.sections.add(-1)          // 통째로 갈렸다
    const found = /^\/sections\/(\d+|-)/.exec(op.path)
    if (found && found[1] !== '-') out.sections.add(Number(found[1]))
    if (op.path.startsWith('/period') || op.path.startsWith('/filters')) out.head = true
    if (op.path.startsWith('/conclusion') || op.path.startsWith('/publishedAt')) out.conclusion = true
  }
  return out
}

File: core/patch.py
"""JSON Patch — 공유 상태를 고치는 유일한 문법.

RFC 6902다. 문서 전체를 다시 보내는 대신 **바뀐 자리만** 보낸다. 이유가 셋이다
(교안 5장 3절).

  · 크기: 섹션이 열 개면 스냅샷은 수 KB고, 패치는 연산 한 줄이다
  · 화면: 무엇이 바뀌었는지 알면 그 자리만 다시 그린다. 스크롤과 포커스가 산다
  · 기록: 패치 배열이 곧 변경 이력이고, 되돌리기도 여기서 나온다

대가는 **기준점이 맞아야 한다**는 것이다. 화면이 가진 문서와 서버가 생각하는
문서가 어긋나면 엉뚱한 자리를 고친다. 그래서 실행을 시작할 때 스냅샷으로
한 번 맞추고 간다.

여섯 연산 중 넷만 쓴다. `copy`와 `test`는 이 랩에 쓸 일이 없다.
"""

from __future__ import annotations

from typing import Any


def replace(path: str, value: Any) -> dict:
    return {"op": "replace", "path": path, "value": value}


def add(path: str, value: Any) -> dict:
    """`/sections/-`는 배열 끝에 붙이라는 RFC 6902의 표기다."""
    return {"op": "add", "path": path, "value": value}


def remove(path: str) -> dict:
    return {"op": "remove", "path": path}


def move(from_: str, path: str) -> dict:
    """배열 안에서 자리를 옮긴다. 화면에서 드래그한 것과 같은 결과다."""
    return {"op": "move", "from": from_, "path": path}


def _segments(path: str) -> list[str]:
    return [s.replace("~1", "/").replace("~0", "~") for s in path.split("/")[1:]]


def _parent(doc: Any, parts: list[str]) -> tuple[Any, str]:
    cursor = doc
    for key in parts[:-1]:
        cursor = cursor[int(key)] if isinstance(cursor, list) else cursor[key]
    return cursor, parts[-1]


def get_at(doc: Any, path: str) -> Any:
    cursor = doc
    for key in _segments(path):
        cursor = cursor[int(key)] if isinstance(cursor, list) else cursor[key]
    return cursor


def apply_one(doc: Any, op: dict) -> None:
    """연산 하나를 제자리에서 적용한다."""
    parts = _segments(op["path"])
    parent, key = _parent(doc, parts)
    kind = op["op"]

    if kind == "add":
        if not isinstance(parent, list):
            parent[key] = op["value"]
        elif key == "-":
            parent.append(op["value"])
        else:
            parent.insert(int(key), op["value"])
    elif kind == "replace":
        if isinstance(parent, list):
            parent[int(key)] = op["value"]
        else:
            parent[key] = op["value"]
    elif kind == "remove":
        if isinstance(parent, list):
            parent.pop(int(key))
        else:
            parent.pop(key, None)
    elif kind == "move":
        value = get_at(doc, op["from"])
        apply_one(doc, {"op": "remove", "path": op["from"]})
        apply_one(doc, {"op": "add", "path": op["path"], "value": value})
    else:
        raise ValueError(f"지원하지 않는 연산: {kind}")


def apply(doc: dict, ops: list[dict]) -> dict:
    """서버도 같은 패치를 자기 사본에 적용한다.

    화면만 고치고 서버가 옛 문서를 들고 있으면, 다음 요청에서 둘이 어긋난
    채로 패치가 나간다. **패치를 만든 쪽도 그것을 적용해야 한다.**
    """
    for op in ops:
        apply_one(doc, op)
    return doc
